grails-app/services/org/pih/warehouse/putaway/PutawayService.groovy

```diff
@@ -30,14 +30,140 @@ import org.pih.warehouse.order.OrderItem
 import org.pih.warehouse.order.OrderItemStatusCode
 import org.pih.warehouse.order.OrderStatus
 import org.pih.warehouse.order.OrderType
+import org.pih.warehouse.product.Product
+import org.pih.warehouse.product.ProductAvailability
+import org.pih.warehouse.inventory.InventoryLevel
+import org.grails.web.json.JSONObject
+import org.pih.warehouse.core.User
 
 @Transactional
 class PutawayService {
 
-    LocationService locationService
-    InventoryService inventoryService
+    def userService
+    def locationService
+    def inventoryService
     def productAvailabilityService
-    GrailsApplication grailsApplication
+    def identifierService
+    def grailsApplication
+
+    void generatePutaways(Location location, User putawayAssignee) {
+        def putawayCandidates = getPutawayCandidates(location)
+        for (PutawayItem putawayCandidate in putawayCandidates) {
+
+
+            // Ignore putaway candidates that already have a putaway order associated with it
+            if (!putawayCandidate.id) {
+
+                log.info "Attempting to generate putaway for putaway candidate ..." + new JSONObject(putawayCandidate?.toJson()).toString(4)
+
+                // Get the next available putaway location based on criteria (available vs unavailable, volume, weight, etc)
+                Location putawayLocation = getNextAvailablePutawayLocation(location, putawayCandidate.product)
+                if (!putawayLocation) {
+                    log.warn("No available putaway location for ${putawayCandidate?.product?.productCode}, lotNumber: ${putawayCandidate?.inventoryItem?.lotNumber?:'Default'}, currentLocation: ${putawayCandidate.currentFacility}.${putawayCandidate?.currentLocation}, quantity=${putawayCandidate.quantity}")
+                    return
+                }
+
+                // Create a new putaway order for the putaway candidate
+                Putaway putaway = new Putaway()
+                putaway.origin = location
+                putaway.destination = location
+                putaway.putawayDate = new Date()
+                putaway.putawayStatus = PutawayStatus.PENDING
+                putaway.putawayAssignee = putawayAssignee
+                putaway.putawayNumber = identifierService.generateOrderIdentifier()
+                putawayCandidate.putawayLocation = putawayLocation
+                putaway.putawayItems.add(putawayCandidate)
+                savePutaway(putaway)
+            }
+        }
+    }
+
+    def getPutawayUsers(Location location) {
+        List<User> users = userService.findUsersByRoleType(RoleType.ROLE_PUTAWAY)
+        if (!users) {
+            String usernameOrId = ConfigurationHolder.config.openboxes.jobs.automaticSlottingJob.defaultPutawayAssignee
```

Comment:
There is this config missing

Reply:
(and job too, but i guess we don't need it for this PR)

Reply:
added